IDE/iotsafe/memory-tls.c

```diff
@@ -0,0 +1,385 @@
+
+/* IoTSAFE Certificate slots */
+
+/* File Slot '03' is pre-provisioned with
+ * local certificate.
+ */
+#define CRT_CLIENT_FILE_ID 0x03     /* pre-provisioned */
+
+/* File Slot '04' is pre-provisioned with the
+ * server's EC public key certificate
+ */
+#define CRT_SERVER_FILE_ID 0x04
+
+/* IoTSAFE Key slots */
+
+/* Key slot '02' is pre-provisioned with
+ * the client private key.
+ */
+#define PRIVKEY_ID      0x02 /* pre-provisioned */
+
+/* Key slot '03' is used by wolfSSL to generate
+ * the ECDH key that will be used during the TLS
+ * session.
+ */
+#define ECDH_KEYPAIR_ID 0x03
+
+/* Key slot '04' is used to store the public key
+ * received from the peer.
+ */
+#define PEER_PUBKEY_ID  0x04
+
+/* Key slot '05' is used to store a public key
+ * used for ecc verification
+ */
+#define PEER_CERT_ID    0x05
+
+/* The following define
+ * activates mutual authentication */
+#define CLIENT_AUTH
+
+
+#define CLIENT_IOTSAFE
+#define CA_ECC
+
+/* client messages to server in memory */
+#define TLS_BUFFERS_SZ (1024 * 8)
+unsigned char to_server[TLS_BUFFERS_SZ];
+int server_bytes;
+int server_write_idx;
+int server_read_idx;
+
+/* server messages to client in memory */
+unsigned char to_client[TLS_BUFFERS_SZ];
+int client_bytes;
+int client_write_idx;
+int client_read_idx;
+
+/* server send callback */
+int ServerSend(WOLFSSL* ssl, char* buf, int sz, void* ctx)
+{
+    if (client_write_idx + sz > TLS_BUFFERS_SZ)
+        return -SSL_ERROR_WANT_WRITE;
+    printf("=== Srv-Cli: %d\n", sz);
+    memcpy(&to_client[client_write_idx], buf, sz);
+    client_write_idx += sz;
+    client_bytes += sz;
+    return sz;
+}
+
+
+/* server recv callback */
+int ServerRecv(WOLFSSL* ssl, char* buf, int sz, void* ctx)
+{
+
+    if (server_bytes - server_read_idx < sz)
+        return -SSL_ERROR_WANT_READ;
+    memcpy(buf, &to_server[server_read_idx], sz);
+    server_read_idx += sz;
+
+    if (server_read_idx == server_write_idx) {
+        server_read_idx = server_write_idx = 0;
+        server_bytes = 0;
+    }
+    printf("=== Srv RX: %d\n", sz);
+    return sz;
+}
+
+
+/* client send callback */
+int ClientSend(WOLFSSL* ssl, char* buf, int sz, void* ctx)
+{
+    if (server_write_idx + sz > TLS_BUFFERS_SZ)
+        return -SSL_ERROR_WANT_WRITE;
+
+    printf("=== Cli->Srv: %d\n", sz);
+    memcpy(&to_server[server_write_idx], buf, sz);
+    server_write_idx += sz;
+    server_bytes += sz;
+
+    return sz;
+}
+
+
+/* client recv callback */
+int ClientRecv(WOLFSSL* ssl, char* buf, int sz, void* ctx)
+{
+
+    if (client_bytes - client_read_idx < sz)
+        return -SSL_ERROR_WANT_READ;
+
+    memcpy(buf, &to_client[client_read_idx], sz);
+    client_read_idx += sz;
+
+    if (client_read_idx == client_write_idx) {
+        client_read_idx = client_write_idx = 0;
+        client_bytes = 0;
+    }
+    printf("=== Cli RX: %d\n", sz);
+    return sz;
+}
+
+static int client_state = 0;
+static int server_state = 0;
+
+uint8_t cert_buffer[2048];
```

Comment:
Does this need to be a global?

Reply:
It is to avoid impacting on the stack size too much (and the size along with it for readability).